### PR TITLE
Start moving forced-id inline to ResourceTable

### PR DESCRIPTION
**Pull request by michaelabuckley**
The join to HFJ_FORCED_ID is expensive.  We are starting to move it into HFJ_RESOURCE.  To allow zero-downtime upgrades, we are doing this in three steps:
1. Add the column and start filling it
2. Add the index, fill any missing values, and start querying the new field.
3. Drop the old table.

This PR implements step 1.

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_3_0/4260-inline-forcedid.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: change
+issue: 4260
+title: "Client assigned resource ids are now stored in a new FORCED_ID column in HFR_RESOURCE.
```
Comment (jamesagnew) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_3_0/4260-inline-forcedid.yaml:4]:
```suggestion
title: "Client assigned resource ids are now stored in a new FORCED_ID column in the HFJ_RESOURCE table.
```

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_3_0/4260-inline-forcedid.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+   This is parallel storage to allow gradual upgrade away from the current HFJ_FORCED_ID join table."
```
Comment (jamesagnew) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_3_0/4260-inline-forcedid.yaml:5]:
```suggestion
   This is parallel storage to allow gradual upgrade away from the current HFJ_FORCED_ID join table, which should improve performance in a future release."
```

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java [hunk 3]
```diff
@@ -544,6 +546,29 @@ public void testCreateDifferentTypesWithSameForcedId() {
 		obs = myObservationDao.read(obsId.toUnqualifiedVersionless(), mySrd);
 	}
 
+	/**
+	 * fixme changelog
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java:550]: Dangling comments?

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java [hunk 4]
```diff
@@ -544,6 +546,29 @@ public void testCreateDifferentTypesWithSameForcedId() {
...
+	 * fixme migration
+	 * We are starting a migration to inline forced_id into RESOURCE_TABLE.
+	 * Verify we are populating the new field AND the old join table for a couple of releases to allow smooth upgrades.
+	 */
+	@Test
+	public void testCreateResourceWithForcedId_populatesResourceTableFieldAndJoinTable() {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java:556]: Recommend also adding a couple tests:

* Test that doesn't use a forced it and verify that the right thing happens (presumably the column remains null?)
* Test that uses serverIdMode of UUID
* Test that uses clientIdMode of ANY with a server-assigned ID
* Test that uses clientIdMode of ANY with a client-assigned numeric ID

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java (hunk 3 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java]: Thanks.

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java (hunk 4 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java]: All the cases now.  And the field is always populated.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java [hunk 5]
```diff
@@ -277,6 +280,65 @@ public class ResourceTable extends BaseHasResource implements Serializable, IBas
...
+					}
+				}
+				Validate.notNull(myFieldIndex, "misconfigured JPA mapping");
+			}
+			return myFieldIndex;
+		}
+	}
+
 	@OneToMany(mappedBy = "myResourceTable", fetch = FetchType.LAZY)
 	private Collection<ResourceHistoryProvenanceEntity> myProvenance;
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java:281]: Do we use the "the" convention for inner class method params?

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java [hunk 6]
```diff
@@ -544,6 +551,80 @@ public void testCreateDifferentTypesWithSameForcedId() {
...
+			"SEQUENTIAL_NUMERIC,ALPHANUMERIC,",
+			"UUID,NOT_ALLOWED,",
+			"UUID,ANY,forcedId",
+			"UUID,ANY,123",
+			"UUID,ANY,", // uuid server-ids still use  hfj_forced_id
+			"UUID,ALPHANUMERIC,forcedId",
+			// illegal "UUID,ALPHANUMERIC,123",
+			"UUID,ALPHANUMERIC,",
+		})
+		public void testCreateResource_populatesResourceTableFhirIdField(
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java:586]: Wow what a great test!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 7]
```diff
@@ -86,6 +86,20 @@ public HapiFhirJpaMigrationTasks(Set<String> theFlags) {
...
+	private void init630() {
+		Builder version = forVersion(VersionEnum.V6_3_0);
+
+		// start forced_id inline migration
+		version
+			.onTable("HFJ_RESOURCE")
+			.addColumn("20221108.1", "FHIR_ID")
+			.nullable()
+			// FHIR ids contain a subset of ascii, limited to 64 chars.
+			.type(ColumnTypeEnum.STRING, 64);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:101]: thought: I cannot wait for this to get fully implemented. Will be so nice to not have to fight the IdHelperService

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java [hunk 8]
```diff
@@ -277,6 +280,65 @@ public class ResourceTable extends BaseHasResource implements Serializable, IBas
...
+		@Override
+		public boolean onPreInsert(PreInsertEvent event) {
+			Object eventEntity = event.getEntity();
+			if (eventEntity instanceof ResourceTable) {
+				ResourceTable resourceTable = (ResourceTable) eventEntity;
+				if (resourceTable.myFhirId == null) {
+					patchFhirIdForInsert(event, resourceTable);
+				}
+			}
+			boolean veto = false;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java:306]: went and read over the interface, good call on making a named variable return value 👍

**Review by tadgh — APPROVED** [review]

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java (hunk 5 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java:281]: We do, Luke.  Thanks.

File: hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java (hunk 6 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-dstu3/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3Test.java:589]: Thanks. The first one skipped all the hard stuff.  In for a penny ...

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java (hunk 8 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceTable.java]: I had to shush IntelliJ.  So bossy.

**Pull request merged**